File: app/fsharp/src/Ch02/TipCalculator.fs
namespace Ch02

/// チップ計算の例
/// グループサイズに応じたチップ率を計算
module TipCalculator =

    /// グループサイズに応じたチップ率を計算
    /// 6人以上 → 20%
    /// 1-5人 → 10%
    /// 0人 → 0%
    let getTipPercentage (names: string list) : int =
        let size = List.length names
        if size > 5 then 20
        elif size > 0 then 10
        else 0

    /// パターンマッチングを使用したチップ率計算
    let getTipPercentageMatch (names: string list) : int =
        match List.length names with
        | n when n > 5 -> 20
        | n when n > 0 -> 10
        | _ -> 0

    /// チップ額を計算
    let calculateTip (billAmount: decimal) (names: string list) : decimal =
        let tipRate = decimal (getTipPercentage names) / 100m
        billAmount * tipRate

    /// 一人当たりの支払額を計算（チップ込み）
    let calculatePerPerson (billAmount: decimal) (names: string list) : decimal option =
        let size = List.length names
        if size = 0 then
            None
        else
            let tip = calculateTip billAmount names
            let total = billAmount + tip
            Some (total / decimal size)

File: app/elixir/lib/ch02/tip_calculator.ex
defmodule Ch02.TipCalculator do
  @moduledoc """
  チップ計算の例

  グループの人数に基づいてチップの割合を計算します。
  純粋関数の例として使用します。
  """

  @doc """
  グループの人数からチップの割合を計算

  - 6人以上: 20%
  - 1-5人: 10%
  - 0人: 0%

  ## Examples

      iex> Ch02.TipCalculator.get_tip_percentage(["Alice", "Bob", "Charlie", "David", "Eve", "Frank"])
      20

      iex> Ch02.TipCalculator.get_tip_percentage(["Alice", "Bob"])
      10

      iex> Ch02.TipCalculator.get_tip_percentage([])
      0

  """
  @spec get_tip_percentage([String.t()]) :: non_neg_integer()
  def get_tip_percentage(names) do
    count = length(names)

    cond do
      count > 5 -> 20
      count > 0 -> 10
      true -> 0
    end
  end

  @doc """
  チップ額を計算

  ## Examples

      iex> Ch02.TipCalculator.calculate_tip(1000, ["Alice", "Bob", "Charlie", "David", "Eve", "Frank"])
      200.0

      iex> Ch02.TipCalculator.calculate_tip(1000, ["Alice", "Bob"])
      100.0

  """
  @spec calculate_tip(number(), [String.t()]) :: float()
  def calculate_tip(bill_amount, names) do
    bill_amount * get_tip_percentage(names) / 100
  end

  @doc """
  チップ込みの合計額を計算

  ## Examples

      iex> Ch02.TipCalculator.calculate_total_with_tip(1000, ["Alice", "Bob"])
      1100.0

  """
  @spec calculate_total_with_tip(number(), [String.t()]) :: float()
  def calculate_total_with_tip(bill_amount, names) do
    bill_amount + calculate_tip(bill_amount, names)
  end

  @doc """
  1人あたりの支払額を計算

  ## Examples

      iex> Ch02.TipCalculator.calculate_per_person(1000, ["Alice", "Bob"])
      550.0

  """
  @spec calculate_per_person(number(), [String.t()]) :: float()
  def calculate_per_person(bill_amount, names) do
    total = calculate_total_with_tip(bill_amount, names)

    case length(names) do
      0 -> 0.0
      n -> total / n
    end
  end

  @doc """
  カスタムチップ率で計算

  ## Examples

      iex> Ch02.TipCalculator.calculate_custom_tip(1000, 15)
      150.0

  """
  @spec calculate_custom_tip(number(), number()) :: float()
  def calculate_custom_tip(bill_amount, tip_percentage) do
    bill_amount * tip_percentage / 100
  end
end
